src/server/active_tran_server.cpp

```diff
@@ -117,4 +119,8 @@ active_tran_server::get_request_handlers ()
 void
 active_tran_server::stop_outgoing_page_server_messages ()
 {
+  for (const auto &sink : m_prior_sender_sink_hooks)
```

Comment:
hmm, this could explain some crashes I was seeing sometime when stopping ATS - some time ago;
however, I have not seen them recently

However, what is the relation between removing the sinks and the transactional log still being produced?
I think the order should be:
- first, the log daemon is stopped so no transactional log can be produced anymore
- secondly, after all prior info has been dispatched to the sinks, the sinks can be removed
Are we sure this is the case?

Reply:
> I think it does matter in the case of graceful stop. It means that log is being legitimately produced but not sent to the Page Servers.

Hmm.. I still doubt it's a problem. Yes, some log records produced legitimately can be lost. However, 
- If the majority of PSes don't receive it, they can be ignored because it's never been committed.
- If the majority of PSes receive it, it will be caught up.

And this is the same to the situation in which a disconnected PS can't receive some legitimately produced log records. I think it's OK in terms of correctness as long as there is no hole among log records. Is it necessary to be careful to stop and wait for all generated logs to be transmitted and clean up resources?

> If the sink is unregistered the link between the synchronous application layer and the asynchronous server-server connection layer is broken and the produced prior log is lost.

I'm not sure I understand you correctly, but this is also saying the above, right?

> Can you please extend the comment such as it reads:

Thanks for your efforts! I'll change this in the next issue as I believe the details might change during the refactoring process with the use of the connection_handler.

Reply:
> Is it necessary to be careful to stop and wait for all generated logs to be transmitted and clean up resources?

Now that you as of it, I see there is no waiting for the termination in the server-server connection layer.
The server-server communication layer is abstracted via the `request_sync_send_queue` and `request_queue_autosend` classes and there is no explicit wait in the dtor of those class for the buffers to be consumed.

As a guide, the closest to a clean stop is done in `page_server::connection_handler::~connection_handler` where:
- incoming communication is stopped
- the already received data is being processed
- the outgoing communication is stopped

> If the sink is unregistered the link between the synchronous application layer and the asynchronous server-server connection layer is broken and the produced prior log is lost.

Take a look at the following call path:
- `logpb_append_prior_lsa_list`
- `prior_sender::send_list`: the prior lists are sent (and copied, unfortunately - this can be improved) - to all sinks (ie: on ATS, to all PS's); still synchronous
  - the sinks in the list are packaged bound calls to `active_tran_server::push_request` (as registered in `active_tran_server::on_boot`); so, still synchronous
- `tran_server::push_request` calls into `request_sync_client_server::push`; still synchronous
- `request_sync_client_server::push` calls into `request_sync_send_queue::push`; and it is only here that the synchronous call chain is broken as the message is pushed on the queue and the sending thread is notified there is new data to dispatch

What I was explaining was a response to your comment:

> The removal sinks just removes the pointer of them from the prior_sender. The sinks are alive in the active_tran_server. It's safe as long as the sinks in active_tran_server are not deleted before handling the prior info requests which are already generated and pushed into the sending queue.

To rephrase: the sinks hold no state; they are a bridge between the application layer and the server-server communication layer (sending thread) and thus, removing them from the log prior sender only breaks this link. Thus, if log would still be produced (it is not, as shown by the exended comment in `xboot_shutdown_server`) the only effect would be that the `prior_sender::send_list` would not send anything. And now I see that the `prior_sender::send_list` implementation is wrong as it reports sending the log although it might send nothing. So, it would be best to add an experimental assert in `prior_sender::send_list` in case log is not being sent at all and see what happens. (I have added http://jira.cubrid.org/browse/LETS-642 for this.)

Reply:
> Now that you mention it, I see there is no waiting for the termination in the server-server connection layer.

OK, now I can see (I guess). 
My opinion was that it's ok in terms of correctness that some legitimately generated log records are lost. However, in normal cases, we don't need to risk losing. We should stop communication gracefully to make sure that everything is under control. 

> As a guide, the closest to a clean stop is done in page_server::connection_handler::~connection_handler where:

Thank you for the guide. Only one missing seems `m_response_broker.terminate()`. Others are done in each dtor. Let's see it in the following issue.

> So, it would be best to add an experimental assert in prior_sender::send_list in case log is not being sent at all and see what happens. (I have added http://jira.cubrid.org/browse/LETS-642 for this.)

I think it's not problematic in terms of correctness to allow the loss of some log records by removal sinks before sending them as long as there is no hole between received log records. But, regardless of the correctness, I assume that you want to make sure that all log records sending requests have been digested properly before removing sinks. Do I understand you correctly? or do we have misunderstandings? I think it is worth your intention since it's more robust and error-prone. I'll consider it in the following issue. Please let me know if there is some misunderstanding.